sips/pending/_posts/2017-01-13-binary-compatibility.md

```diff
@@ -0,0 +1,168 @@
+
+  - public fields. Can be simulated by explicitly defining public getters and setters that access a private field;
+  - lazy vals. Can be simulated by explicitly writing an implementation in source;
+  - case classes. Can be simulated by explicitly defining getters and other members synthesized for a case class(`copy`, `productArity`, `apply`, `unApply`, `unapply`).
+
+The features listed below cannot be easily re-implemented in a class or trait annotated with `@binaryCompatible`.
+  - default arguments;
+  - default methods. See Addendum;
+  - constant types(both explicit and inferred);
+  - inline.
+    
+## `@binaryCompatible` and Scala.js
+
+Allowing to write API-defining classes in Scala instead of Java will allow them to compile with Scala.js, 
+which would have benefit of sharing the same source for two ecosystems.
+
+Scala.js currently is binary compatible as long as original bytecode compiled by Scala JVM is binary compatible.
+Providing stronger binary compatibility guarantees for JVM will automatically provide stronger guarantees for Scala.js.
+ 
+
+## Comparison with MiMa ##
+The Migration Manager for Scala (MiMa in short) is a tool for diagnosing binary incompatibilities for Scala libraries.
+MiMa allows to compare binary APIs of two already compiled classfiles and reports errors if APIs do not match perfectly.
+
+MiMa and `@binaryCompatible` complement each other, as `@binaryCompatible` helps to develop APIs that stay compatible 
+across major versions, while MiMa checks that previously published artifacts indeed have the same API.
+
+`@binaryCompatible` does not compare the currently compiled class or trait against previous version, 
+so introduction of new members won't be prohibited. This is a use-case for MiMa.
+  
+MiMa does not indicate how hard, if possible, would it be to maintain compatibility of a class across future versions of Scala.
+Multiple features of Scala, most notably lazy vals and traits, has been compiled diffently by different Scala versions
+making porting existing compiled bytecode across versions very hard. 
+MiMa will complain retroactively that the new version is incompatible with the old one. 
+`@binaryCompatible` will instead indicate at compile time that the old version had used features whose encoding is prone to change.
+This provides early guidance and warning when designing long-living APIs before they are publicly released.  
+
+## Compilation scheme ##
+No modification of typer or any existing phase is planned. The current proposed scheme introduces a late phase that runs before the very bytecode emission that checks that:
+ - classes and traits annotated as  `@binaryCompatible` are on the top level; 
+ - no non-private members where introduced inside classes and traits annotated as  `@binaryCompatible` by compiler using phase travel;
+ - no non-private members inside classes and traits annotated as  `@binaryCompatible` has changed their signature from the one written by developer.
```

Comment:
What does this mean? Does erasure of type arguments count as a change, for instance?